Pr add haskell compiler type

mberkowitz opened the pull request:
Have haskell-compile choose a build tool _(stack, cabal, ghc)_ the same way that
haskell-interactive mode chooses a repl _(stack repl, cabal repl, ghci)_.

That is, add a new customization variable _haskell-compiler-type_ analogous to
_haskell-process-type_. It's an enum with values to select stack, cabal, or ghc;
and the value 'auto tells emacs to find the closest stack.yaml or *.cabal file
to decide.

Also drops the value 'cabal-new-repl for haskell-process-type, since cabal now
defaults to the new (v2) behavior. I did not add a way to choose cabal v1.

The same code (factored out as (haskell-build-type)) is used for compiler and
repl. It does a rough guess, which is good enough since the chosen command can
be previewed and edited.

haskell-compiler-type replaces the bool haskell-compile-ignore-cabal, which
was biased against stack, and inconsistent with haskell-interactive-mode.

Why two variables instead of one? I can't imagine wanting to compile and
interact with different tools, but I didn't want to make haskell-compile
depend on haskell-interactive-mode, since a user may prefer a different emacs
library instead of either of them.

This pull request is similar in intent to the long-pending #1545,
but is based on a more recent haskell-compile.

purcell commented on haskell-customize.el:
```diff
@@ -429,42 +417,48 @@ imports."
   "The path which is considered as project root, this is determined by the
 presence of a *.cabal file or stack.yaml file or something similar.")
 
+(defun haskell-build-type ()
```
Just an observation that there are now many instances of haskell-related elisp code that do something like the below, e.g. in `dante`. I wish we could unify them!

purcell reviewed (commented):
Mostly looks great afaict, and I'm inclined towards merging this relatively aggressively, but see my notes about removal of the `cabal-new-repl` option and potential breakage for users.

mberkowitz commented on haskell-customize.el:
```diff
@@ -66,9 +67,8 @@ Otherwise if there's a *.cabal file, cabal-repl will be used.
                  (const ghci)
-                 (const cabal-repl)
                  (const stack-ghci)
-                 (const cabal-new-repl))
```
Seems reasonable to keep the value _'cabal-new-repl_ and emit _"cabal new-xxx"_
What about supporting _"cabal v1-xxx"_ and _"cabal v2-xxx"?_  My preference is don't, as it makes the elisp code too complicated, and it is so easy for the user to edit the generated command before it is run.
In general, is it worth detecting and coping with the user's version of _cabal_?

purcell replied:
Not worth detecting, and I'm fine with only one version of "cabal repl" being supported, ie. I don't think we need to make it switchable between "repl" and "new-repl". However, what I'm wary of is that somebody has configured a setting that then silently has no effect. I would rather they get a hard error, or have the `cabal-new-repl` value treated as `cabal-repl`, perhaps with a warning message.

mberkowitz replied:
OK. I will restore  _'cabal-new_ and _'cabal-new-repl_, which will emit _cabal new-build ..._ and _cabal new-repl ..._,
as _cabal_ still supports them. Perhaps they'll be needed again when v3 is introduced.

mberkowitz replied:
Did that, lbut it ooks too complicated. After reading your comments again,  I agree with not emitting _cabal new-repl_ or _cabal new-build_, and just accepting the obsolete customization enum 'cabal-new-repl with a warning.
I will clean up my branch -- in fact, collapse it into one commit. I hope that's ok, this is my 1st pull request.

purcell replied:
> I will clean up my branch -- in fact, collapse it into one commit. I hope that's ok, this is my 1st pull request.

It's not super important for it to be a single commit, tbh, because I can easily squash it when merging if necessary. But a bit of squashing on your end just to keep the commits clean is of course fine, and helps preserve the history.

mberkowitz commented:
Please review again.
Rewritten to allow obsolete value haskell-process-type = 'cabal-new-repl,
converted to the equivalent value 'cabal-repl (with a warning) on 1st use.
Also fixed a couple of bugs found hand-testing haskell-compiler-type.
I don't understand the CI test results on my branch. When I test by hand (_make check_), I see 3 errors in files I haven't changed:
Ran 539 tests, 536 results as expected, 1 unexpected, 2 skipped (2020-09-14 20:32:36-0700, 17.627873 sec)
121 expected failures

1 unexpected results:
   FAILED  haskell-indent-put-region-in-literate-2

2 skipped results:
  SKIPPED  haskell-indent-in-comment-1
  SKIPPED  haskell-indentation-altj-comment

purcell commented:
The CI results are spurious, sorry: fixed in a71bc2bc204edb9cc30167587a773de40d8a2bf6.

This all looks good to me, so I'll go ahead and merge, and if any related issues come up here then I'll tag you in and we can figure them out together. Thanks again!

mberkowitz commented:
Thank you so much.

On Tue, Sep 15, 2020 at 3:20 PM Steve Purcell <notifications@github.com>
wrote:

>
>
> The CI results are spurious, sorry: fixed in a71bc2b
> <https://github.com/haskell/haskell-mode/commit/a71bc2bc204edb9cc30167587a773de40d8a2bf6>
> .
>
>
> This all looks good to me, so I'll go ahead and merge, and if any related
> issues come up here then I'll tag you in and we can figure them out
> together. Thanks again!
>
>
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/haskell/haskell-mode/pull/1717#issuecomment-693010059>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHRT75OE6EJEBVIG7QM5ADSF7SDBANCNFSM4RAQ65RA>
> .
>
>
>
